File: python/241.py

class Solution(object):
    def diffWaysToCompute(self, input):
        """
        :type input: str
        :rtype: List[int]
        """
        if input.isdigit():
            return [int(input)]
        res = []
        op = {'+': lambda a, b: a+b,
              '-': lambda a, b: a-b,
              '*': lambda a, b: a*b}
        for i, c in enumerate(input):
            if c in '+-*':
                lefts = self.diffWaysToCompute(input[:i])
                rights = self.diffWaysToCompute(input[i+1:])
                res.extend([op[c](l, r) for l in lefts for r in rights])
        return res


sol = Solution()
print(sol.diffWaysToCompute("15-7*6+24"))

File: cpp/241.cpp
#include <string>
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  vector<int> diffWaysToCompute(string input) {
    bool alldigit = true;
    int val = 0;
    for (char c : input) {
      if (c < '0' || c > '9') {alldigit = false; break;}
      val = val * 10 + c - '0';
    }
    if (alldigit) return {val};

    vector<int> ans;
    for (size_t i = 0; i < input.size(); i++) {
      if (input[i] == '+' || input[i] == '-' || input[i] == '*') {
        vector<int> lefts = diffWaysToCompute(input.substr(0, i));
        vector<int> rights = diffWaysToCompute(input.substr(i + 1, input.size() - i - 1));
        for (int l : lefts)
          for (int r : rights)
            switch (input[i]) {
            case '+': ans.push_back(l + r); break;
            case '-': ans.push_back(l - r); break;
            case '*': ans.push_back(l * r); break;
            }
      }
    }
    return ans;
  }
};

int main() {
  for (int i : Solution().diffWaysToCompute("15-5"))
    cout << i << " ";
  cout << endl;
}
